packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmObjectHelper.kt

```diff
@@ -153,45 +153,59 @@ internal object RealmObjectHelper {
-        elementType: KClass<*>,
+        propertyMetadata: PropertyMetadata,
+        elementType: KClass<R & Any>,
         operatorType: CollectionOperatorType
     ): ManagedRealmList<R> {
-        val listPtr = RealmInterop.realm_get_list(obj.objectPointer, key)
+        val listPtr = RealmInterop.realm_get_list(obj.objectPointer, propertyMetadata.key)
         val operator =
-            createListOperator<R>(listPtr, elementType, obj.mediator, obj.owner, operatorType)
+            createListOperator<R>(listPtr, elementType, propertyMetadata, obj.mediator, obj.owner, operatorType)
         return ManagedRealmList(listPtr, operator)
     }
 
     @Suppress("LongParameterList")
     private fun <R> createListOperator(
         listPtr: RealmListPointer,
-        clazz: KClass<*>,
+        clazz: KClass<R & Any>,
+        propertyMetadata: PropertyMetadata,
         mediator: Mediator,
         realm: RealmReference,
         operatorType: CollectionOperatorType
     ): ListOperator<R> {
         val converter: RealmValueConverter<R> =
-            converter<Any>(clazz, mediator, realm) as CompositeConverter<R, *>
+            converter<R>(clazz, mediator, realm) as CompositeConverter<R, *>
         return when (operatorType) {
             CollectionOperatorType.PRIMITIVE ->
                 PrimitiveListOperator(mediator, realm, listPtr, converter)
-            CollectionOperatorType.REALM_OBJECT ->
-                RealmObjectListOperator(mediator, realm, listPtr, clazz, converter)
-            CollectionOperatorType.EMBEDDED_OBJECT -> EmbeddedRealmObjectListOperator(
-                mediator,
-                realm,
-                listPtr,
-                clazz,
-                converter as RealmValueConverter<EmbeddedRealmObject>
-            ) as ListOperator<R>
+            CollectionOperatorType.REALM_OBJECT -> {
+                val classKey: ClassKey = realm.schemaMetadata.getOrThrow(propertyMetadata.linkTarget).classKey
+                RealmObjectListOperator(
+                    mediator,
+                    realm,
+                    listPtr,
+                    clazz,
+                    classKey,
+                    converter
+                )
+            }
+            CollectionOperatorType.EMBEDDED_OBJECT -> {
+                val classKey: ClassKey = realm.schemaMetadata.getOrThrow(propertyMetadata.linkTarget).classKey
+                EmbeddedRealmObjectListOperator(
+                    mediator,
+                    realm,
+                    listPtr,
+                    clazz as KClass<EmbeddedRealmObject>,
+                    classKey,
+                    converter as RealmValueConverter<EmbeddedRealmObject>
+                ) as ListOperator<R>
+            }
         }
     }
 
     internal inline fun <reified R : Any> getSet(
         obj: RealmObjectReference<out BaseRealmObject>,
         propertyName: String
-    ): ManagedRealmSet<Any?> {
+    ): ManagedRealmSet<R?> {
```

Comment:
Does this comment https://github.com/realm/realm-kotlin/blob/324a4f69ea272b96d6337c888e079e57a0ee9c54/packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/RealmObjectHelper.kt#L68 not apply anymore? I haven't been able to use generics on native in the `realm_value_t` pull request.

Reply:
The comment is really old, so I just thought I would try to use generics properly the places that I had to touch. It seems to work for most of it, but haven't gone through all the cases relevant to this comment, so didn't remove the comment just yet.